File: sdk/python/bitcoin_bastion_sdk/safety.py
from __future__ import annotations

from collections.abc import Mapping, Sequence

from bitcoin_bastion_sdk.errors import BitcoinWalletSecretForbiddenError

SAFETY_MESSAGE = (
    "Never submit seed phrases, private keys, wallet files, xprv/yprv/zprv, "
    "or signing material to Bitcoin Bastion."
)
FORBIDDEN_INDICATORS = (
    "seed phrase",
    "mnemonic",
    "private key",
    "xprv",
    "yprv",
    "zprv",
    "wallet.dat",
    "keystore",
    "12 words",
    "24 words",
    "signing material",
    "wif",
)
FORBIDDEN_KEYS = frozenset(
    {"seed", "seed_phrase", "mnemonic", "wallet_seed", "bitcoin_seed", "private_key", "xprv", "wif"}
)


def assert_safe(value: object) -> None:
    if _contains_sensitive(value):
        raise BitcoinWalletSecretForbiddenError()


def _contains_sensitive(value: object) -> bool:
    if isinstance(value, str):
        lowered = value.casefold()
        return any(term in lowered for term in FORBIDDEN_INDICATORS)
    if isinstance(value, Mapping):
        return any(
            str(key).casefold() in FORBIDDEN_KEYS
            or _contains_sensitive(str(key))
            or _contains_sensitive(item)
            for key, item in value.items()
        )
    if isinstance(value, Sequence) and not isinstance(value, (bytes, bytearray, str)):
        return any(_contains_sensitive(item) for item in value)
    return False

File: sdk/typescript/src/safety.ts
import { BastionSafetyError } from "./errors.js";

const SENSITIVE_PATTERNS = [
  "seed phrase",
  "mnemonic",
  "private key",
  "bitcoin seed",
  "wallet seed",
  "xprv",
  "yprv",
  "zprv",
  "wallet.dat",
  "keystore",
  "12 words",
  "24 words",
  "signing material",
  "bitcoin_seed",
  "private_key",
];

export const SAFETY_MESSAGE =
  "Never submit seed phrases, private keys, wallet files, xprv/yprv/zprv, or signing material to Bitcoin Bastion.";
export const WALLET_SECRET_REJECTION_MESSAGE = "Bitcoin Bastion does not require your Bitcoin wallet seed or private key.";

const FORBIDDEN_AUTH_FIELDS = /^(seed|seedphrase|mnemonic|privatekey|walletprivatekey|xprv|bitcoinseed|lightningseed)$/i;
export function assertNoWalletSecretFields(value: unknown): void {
  if (value && typeof value === "object") for (const [key, item] of Object.entries(value)) {
    if (FORBIDDEN_AUTH_FIELDS.test(key.replace(/[_-]/g, ""))) throw new BastionSafetyError(WALLET_SECRET_REJECTION_MESSAGE);
    assertNoWalletSecretFields(item);
  }
}

export function containsSensitiveMaterial(value: unknown): boolean {
  return SENSITIVE_PATTERNS.some((pattern) => flatten(value).toLowerCase().includes(pattern));
}

export function assertNoSensitiveMaterial(value: unknown): void {
  if (containsSensitiveMaterial(value)) throw new BastionSafetyError(SAFETY_MESSAGE);
}

function flatten(value: unknown): string {
  if (typeof value === "string") return value;
  if (Array.isArray(value)) return value.map(flatten).join(" ");
  if (value && typeof value === "object") return Object.entries(value).map(([key, item]) => `${key} ${flatten(item)}`).join(" ");
  return String(value ?? "");
}
